### PR TITLE
perf(*): eliminate uneccessary JSON marshalling

### LINKED ISSUE #3423: Protobuf util intermediary JSON objects consuming excessive memory.
### What happened?

In pkg/util/proto/proto.go, converting protobufs to maps and back creates intermediate JSON object:

```
func FromMap(in map[string]interface{}, out proto.Message) error {
        content, err := json.Marshal(in)
        if err != nil {
                return err
        }
        return FromJSON(content, out)
}
```

This is consuming excessive amounts of memory.

Should marshal to and from protobufs more directly.

### PR DESCRIPTION
**Pull request by parkanzky**
### Summary

Use protobufs for resource `Spec` directly, rather than marshaling in and out of `map[string]interface{}` via intermediate JSON object, saving a ton of memory. 

Also moved the protobuf access helpers in `pkg/plugins/resources/k8s/native/api/v1alpha1/` into generated code. There is probably room for more generation in `api/mesh/v1alpha1/`, but leaving that out in the interest of time.

### Full changelog

* Fix memory consumption of resource objects.

### Issues resolved

Fix #3423

### Testing

- [x] Unit tests
- [x] E2E tests
- [x] Manual testing on Universal
- [x] Manual testing on Kubernetes

### Backwards compatibility

- [ ] Update [`UPGRADE.md`](UPGRADE.md) with any steps users will need to take
      when upgrading.
- [x] Add `backport-to-stable` label if the code is backwards compatible. Otherwise, list breaking changes.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -21,7 +21,7 @@ require (
 	github.com/golang/protobuf v1.5.2
 	github.com/google/uuid v1.3.0
 	github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
-	github.com/gruntwork-io/terratest v0.38.5
+	github.com/gruntwork-io/terratest v0.38.6
```
Comment (jpeach) [inline go.mod:24]: Unless the upgrade is needed for this PR, I'd recommend pulling it out to reduce the scope and to make the update more easily cherry-pickable.

File: pkg/plugins/runtime/gateway/match/policy.go [hunk 2]
```diff
@@ -6,6 +6,7 @@ import (
 	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
 	"github.com/kumahq/kuma/pkg/core/policy"
 	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/gateway/route"
```
Comment (jpeach) [inline pkg/plugins/runtime/gateway/match/policy.go:9]: Can you rebase this PR so that all the unrelated changes go away?

File: tools/resource-gen/main.go [hunk 3]
```diff
@@ -19,6 +19,126 @@ import (
 	_ "github.com/kumahq/kuma/api/mesh/v1alpha1"
 )
 
+// NativeApuTemplate for creating a Kuma resource.
```
Comment (jpeach) [inline tools/resource-gen/main.go:22]: s/NativeApuTemplate/NativeApiTemplate/

File: tools/resource-gen/main.go [hunk 4]
```diff
@@ -19,6 +19,126 @@ import (
...
+var NativeApiTemplate = template.Must(template.New("resource").Parse(`
+// Generated by tools/resource-gen
+// Run "make generate" to update this file.
+
+{{ $pkg := printf "%s_proto" .Package }}
+{{ $tk := "` + "`" + `" }}
```
Comment (jpeach) [inline tools/resource-gen/main.go:28]: This is making my head hurt .. what's going on here?

**Review by jpeach — DISMISSED** [review]
I didn't have time to finish reviewing yet, but please rebase onto latest master to remove a lot of the unrelated. You'll also need to update the PR and commit subject to match the contributing standards.

File: tools/resource-gen/main.go (hunk 4 above)
Comment (michaelbeaumont) [reply tools/resource-gen/main.go:28]: [It's impossible to escape backticks inside a string literal](https://github.com/golang/go/issues/24475) + Go relying so heavily on writing code generation via string templating :shrug: .
Seems like a good way to solve this.

Comment (jpeach) [reply tools/resource-gen/main.go:28]: OK, I see the first "`" closes the string that started at `Parse`, then we concat a double-quotes backtick, then we concat the remainder of the parse string.

I think it would be simpler to add a `.Backtick` member or function to the template context :)

File: tools/resource-gen/main.go [hunk 5]
```diff
@@ -19,6 +19,126 @@ import (
...
+
+{{range .Resources}}
+{{- if not .SkipNativeGenerate }}
+
+// +kubebuilder:object:root=true
+type {{.ResourceType}} struct {
+	metav1.TypeMeta   {{ $tk }}json:",inline"{{ $tk }}
+	metav1.ObjectMeta {{ $tk }}json:"metadata,omitempty"{{ $tk }}
+
+	Mesh   string                        {{ $tk }}json:"mesh,omitempty"{{ $tk }}
```
Comment (jpeach) [inline tools/resource-gen/main.go:52]: IIUC, types where `.Global` is true should not have a mesh field.

File: tools/resource-gen/main.go [hunk 6]
```diff
@@ -184,32 +304,36 @@ func SelectorsForMessage(m protoreflect.MessageDescriptor) []string {
...
-	KdsDirection     string
+	ResourceName       string
+	ResourceType       string
+	ProtoType          string
+	Selectors          []string
+	SkipRegistration   bool
+	SkipValidation     bool
+	SkipNativeGenerate bool
+	ScopeNamespace     bool
+	Global             bool
```
Comment (lahabana) [inline tools/resource-gen/main.go:315]: As @jpeach pointed out Global and ScopeNamespace are the same thing I believe.

File: api/mesh/v1alpha1/timeout_helpers.go [hunk 7]
```diff
@@ -12,3 +17,22 @@ func (x *Timeout_Conf) GetConnectTimeoutOrDefault(defaultConnectTimeout time.Dur
...
+	proto.Merge(out, t)
+}
+func (t *Timeout) DeepCopy() *Timeout {
+	if t == nil {
+		return nil
+	}
+	out := new(Timeout)
+	t.DeepCopyInto(out)
+	return out
+}
```
Comment (lahabana) [inline api/mesh/v1alpha1/timeout_helpers.go:38]: Can't these all be auto generated too? Sounds like it's a bunch of boilerplate code that could just autogenerated in a file.

File: tools/resource-gen/main.go (hunk 6 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:315]: I tried that originally, and only added the extra option when results didn't match. Eg Dataplane insight is ScopeNamespace, but not Global protobuf option.

File: api/mesh/v1alpha1/timeout_helpers.go (hunk 7 above)
Comment (parkanzky) [reply api/mesh/v1alpha1/timeout_helpers.go:38]: @lahabana They can. This is what I was referring to in the summary with "There is probably room for more generation in api/mesh/v1alpha1/". These were really easy to append to the existing files, but we can put them under automation if that is what we want to spend our time on.

File: tools/resource-gen/main.go (hunk 4 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:28]: @jpeach I think pulling it out into a part of the code where the escaping is more readable is a good idea. I'm going to copy it back into short-named var like `tk` in the template though because `{{.Backtick}}` is just a lot to put into every place where a backtick usually goes.

File: tools/resource-gen/main.go (hunk 5 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:52]: @jpeach MeshInsight has `option (kuma.mesh.resource).global = true;` and a `Mesh` field.

I'll see if I can make this match with the existing options. Maybe `option (kuma.mesh.resource).package = "mesh";`

File: tools/resource-gen/main.go (hunk 6 above)
Comment (lahabana) [reply tools/resource-gen/main.go:315]: That's weird :) Could it be a bug?

File: api/mesh/v1alpha1/timeout_helpers.go (hunk 7 above)
Comment (lahabana) [reply api/mesh/v1alpha1/timeout_helpers.go:38]: I guess now that it's done it's ok. It's just easier to generate new policies if everything is auto gen :)

File: tools/resource-gen/main.go (hunk 6 above)
Comment (lahabana) [reply tools/resource-gen/main.go:315]: Actually I'm talking crap these are different :)

File: tools/resource-gen/main.go (hunk 5 above)
Comment (lahabana) [reply tools/resource-gen/main.go:52]: we should also open an issue saying that MeshInsights shouldn't have a Mesh field.

File: tools/resource-gen/main.go (hunk 4 above)
Comment (lahabana) [reply tools/resource-gen/main.go:28]: Could we just work around this by having all templates in a folder as raw text and use goembed to load them?

File: api/mesh/v1alpha1/timeout_helpers.go (hunk 7 above)
Comment (michaelbeaumont) [reply api/mesh/v1alpha1/timeout_helpers.go:38]: Will something fail to compile somewhere if someone forgets to add the custom Un/MarshalJSON methods?

Comment (parkanzky) [reply api/mesh/v1alpha1/timeout_helpers.go:38]: Since this PR is a high priority for the next release which will be cut soon, I propose we move this code under auto-gen as part of this PR only if there is time after fixing all the tests and other necessary fallout from this change.

File: go.mod (hunk 1 above)
Comment (parkanzky) [reply go.mod:24]: Something is up with my dev env and `make generate` which causes this. I usually catch it and revert prior to PR. I'll look into root cause soon.

File: tools/resource-gen/main.go (hunk 5 above)
Comment (jpeach) [reply tools/resource-gen/main.go:52]: IIUC, all the Kubernetes types  have a Mesh field so they can implement the `KubernetesObject` interface.

File: api/mesh/v1alpha1/timeout_helpers.go (hunk 7 above)
Comment (lahabana) [reply api/mesh/v1alpha1/timeout_helpers.go:38]: Agreed with @parkanzky (it's ok to open an issue to do this after if no time on the PR).
@michaelbeaumont yes it will though it's annoying to have to fight against the compiler when creating things no?

File: tools/resource-gen/main.go (hunk 5 above)
Comment (lahabana) [reply tools/resource-gen/main.go:52]: Right but for MeshInsights we can add this method ourselves as a helper like we do for Mesh (and return/set Name).

File: tools/resource-gen/main.go [hunk 8]
```diff
@@ -19,6 +19,126 @@ import (
...
+func (cb *{{.ResourceType}}) GetMesh() string {
+	return cb.Mesh
+}
+
+func (cb *{{.ResourceType}}) SetMesh(mesh string) {
+	cb.Mesh = mesh
+}
+
+func (cb *{{.ResourceType}}) GetSpec() proto.Message {
+{{- if eq .ResourceType "DataplaneInsight" }}
```
Comment (jpeach) [inline tools/resource-gen/main.go:91]: Instead of special-casing DataplaneInsight, how about just skipping helper generation for this type?

File: tools/resource-gen/main.go [hunk 9]
```diff
@@ -19,6 +19,126 @@ import (
...
+}
+
+func (cb *{{.ResourceType}}) Scope() model.Scope {
+{{- if .ScopeNamespace }}
+	return model.ScopeNamespace
+{{- else }}
+	return model.ScopeCluster
+{{- end}}
+}
+
```
Comment (jpeach) [inline tools/resource-gen/main.go:113]: Why don't you also generate the JSON marshaller and deep copy implementations?

File: tools/resource-gen/main.go [hunk 10]
```diff
@@ -19,6 +19,126 @@ import (
...
+// +kubebuilder:object:root=true
+type {{.ResourceType}} struct {
+	metav1.TypeMeta   {{ $tk }}json:",inline"{{ $tk }}
+	metav1.ObjectMeta {{ $tk }}json:"metadata,omitempty"{{ $tk }}
+
+	Mesh   string                        {{ $tk }}json:"mesh,omitempty"{{ $tk }}
+{{- if eq .ResourceType "DataplaneInsight" }}
+	Status   *{{ $pkg }}.{{.ProtoType}} {{ $tk }}json:"status,omitempty"{{ $tk }}
+{{- else}}
+	Spec   *{{ $pkg }}.{{.ProtoType}} {{ $tk }}json:"spec,omitempty"{{ $tk }}
```
Comment (jpeach) [inline tools/resource-gen/main.go:56]: What does crd-gen do with this?

File: pkg/test/apis/sample/v1alpha1/sample_types_helpers.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+func (m *TrafficRoute) UnmarshalJSON(data []byte) error {
+	return util_proto.FromJSON(data, m)
+}
+
+func (m *TrafficRoute) MarshalJSON() ([]byte, error) {
+	return util_proto.ToJSON(m)
+}
+func (t *TrafficRoute) DeepCopyInto(out *TrafficRoute) {
+	proto.Merge(out, t)
```
Comment (jpeach) [inline pkg/test/apis/sample/v1alpha1/sample_types_helpers.go:17]: This should be `util_proto.Merge()`.

File: tools/resource-gen/main.go [hunk 12]
```diff
@@ -19,6 +19,126 @@ import (
...
+{{- end}}
+}
+
+// +kubebuilder:object:root=true
+type {{.ResourceType}}List struct {
+	metav1.TypeMeta {{ $tk }}json:",inline"{{ $tk }}
+	metav1.ListMeta {{ $tk }}json:"metadata,omitempty"{{ $tk }}
+	Items           []{{.ProtoType}} {{ $tk }}json:"items"{{ $tk }}
+}
```
Comment (jpeach) [inline tools/resource-gen/main.go:65]: I'm not sure that I'm convinced that generating the CRD is worth it. It makes it harder to add kubebuilder annotations that might be needed, and to add doc comments that will end up in the CRD.

I do agree with generating the helpers though.

File: api/mesh/options.proto [hunk 13]
```diff
@@ -27,6 +27,12 @@ message KumaResourceOptions {
 
   KumaKdsOptions kds = 10;
   KumaWsOptions ws = 7;
+
+  // Whether scope is "Namespace"; Otherwise to "Cluster".
+  bool scope_namespace = 11;
```
Comment (jpeach) [inline api/mesh/options.proto:32]: Is there a relationship between kuma scope and kubernetes scope? For example, do we have a rule that global kuma scope objects are cluster kubernetes scope?

File: api/mesh/options.proto [hunk 14]
```diff
@@ -27,6 +27,12 @@ message KumaResourceOptions {
...
+
+  // Whether to skip generation of native API helper functions.
+  bool skip_native_generate = 12;
```
Comment (jpeach) [inline api/mesh/options.proto:35]: Maybe "skip_kubernetes_wrappers" or something is a more obvious name?

File: api/mesh/v1alpha1/circuit_breaker_helpers.go [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+}
+func (t *CircuitBreaker) DeepCopyInto(out *CircuitBreaker) {
+	proto.Merge(out, t)
+}
+func (t *CircuitBreaker) DeepCopy() *CircuitBreaker {
+	if t == nil {
+		return nil
+	}
+	out := new(CircuitBreaker)
+	t.DeepCopyInto(out)
```
Comment (jpeach) [inline api/mesh/v1alpha1/circuit_breaker_helpers.go:24]: Can you do `proto.Clone()` here?

File: api/mesh/v1alpha1/circuit_breaker_helpers.go [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+func (m *CircuitBreaker) UnmarshalJSON(data []byte) error {
+	return util_proto.FromJSON(data, m)
+}
+
+func (m *CircuitBreaker) MarshalJSON() ([]byte, error) {
+	return util_proto.ToJSON(m)
+}
+func (t *CircuitBreaker) DeepCopyInto(out *CircuitBreaker) {
+	proto.Merge(out, t)
```
Comment (jpeach) [inline api/mesh/v1alpha1/circuit_breaker_helpers.go:17]: `proto.Merge` takes `Message` parameters, but here you are passing the full CRD object? Do you need to deep copy the meta and other fields before copying the protobuf?

**Review by jpeach — COMMENTED** [review]
There's a lot here, and it's not super clear to me which parts are absolutely necessary for the performance issue, and which parts are nice-to-have refactorings. I'm a bit concerned that exposing the proto types directly in the CRD will prevent us adopting crd-gen.

File: tools/resource-gen/main.go (hunk 4 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:28]: Going to do more with templates in a subsequent PR - autogen the rest of the helpers, add new autogen type, move templates into their own files so that backticks don't need special handling. Will open issues if I don't get to it today.

File: tools/resource-gen/main.go (hunk 5 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:52]: Preserving existing behavior in this PR, will open an issue.

File: api/mesh/v1alpha1/timeout_helpers.go (hunk 7 above)
Comment (parkanzky) [reply api/mesh/v1alpha1/timeout_helpers.go:37]: Mentioned above, will open a subsequent PR with template suggestions and open an issue if I don't get to that today.

File: api/mesh/v1alpha1/circuit_breaker_helpers.go (hunk 16 above)
Comment (parkanzky) [reply api/mesh/v1alpha1/circuit_breaker_helpers.go]: In this scope, `CircuitBreaker` is the `Message`

File: tools/resource-gen/main.go [hunk 17]
```diff
@@ -296,11 +422,13 @@ func main() {
 		switch gen {
 		case "type":
```
[the remark this answers is not in the document]
Comment (parkanzky) [reply tools/resource-gen/main.go:428]: Will do this in the next PR or open an issue.

File: tools/resource-gen/main.go (hunk 8 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:91]: More template wrangling in the next PR. I think we actually need an issue to determine if this field is actually correct or if it should be `Spec` like everything else.

File: tools/resource-gen/main.go (hunk 10 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:56]: I don't know, but for now preserving existing behavior. Leaving this hack in place until we determine if it should just be changed. Opening an issue on that.

File: tools/resource-gen/main.go (hunk 9 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:113]: Will add that in a subsequent PR.

File: api/mesh/options.proto (hunk 13 above)
Comment (parkanzky) [reply api/mesh/options.proto:32]: Not that I could discern. I tried using the global to identify namespace and got results that differed in autogen from existing code.

File: tools/resource-gen/main.go (hunk 12 above)
Comment (parkanzky) [reply tools/resource-gen/main.go:65]: Going to leave the CRD generation in for now. It's easy enough to just remove that autogen and split up the previously autogened code back into individual files if we want in the future. Will generate the helpers in a subsequent PR.

File: api/mesh/v1alpha1/circuit_breaker_helpers.go (hunk 15 above)
Comment (parkanzky) [reply api/mesh/v1alpha1/circuit_breaker_helpers.go:22]: I would assume so. I was following the example of the controller autogen deepcopy files which always call `DeepCopyInto` from `DeepCopy`. Plan is to generate the helpers next, so makes sense to change it there instead of manually updating all these individual files.

**Review by michaelbeaumont — APPROVED** [review]
LGTM

**Comment by parkanzky** [comment]
Added the following issues that came up in code review comments -

https://github.com/kumahq/kuma/issues/3477
https://github.com/kumahq/kuma/issues/3479
https://github.com/kumahq/kuma/issues/3480
https://github.com/kumahq/kuma/issues/3481
